pkg/ttnpb/application.pb.validate.go

```diff
@@ -0,0 +1,1005 @@
+// Code generated by protoc-gen-fieldmask. DO NOT EDIT.
+
+package ttnpb
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"net"
+	"net/mail"
+	"net/url"
+	"regexp"
+	"strings"
+	"time"
+	"unicode/utf8"
+
+	"github.com/gogo/protobuf/types"
+)
+
+// ensure the imports are used
+var (
+	_ = bytes.MinRead
+	_ = errors.New("")
+	_ = fmt.Print
+	_ = utf8.UTFMax
+	_ = (*regexp.Regexp)(nil)
+	_ = (*strings.Reader)(nil)
+	_ = net.IPv4len
+	_ = time.Duration(0)
+	_ = (*url.URL)(nil)
+	_ = (*mail.Address)(nil)
+	_ = types.DynamicAny{}
+)
+
+// ValidateFields checks the field values on Application with the rules defined
+// in the proto definition for this message. If any rules are violated, an
+// error is returned.
+func (m *Application) ValidateFields(paths ...string) error {
+	if m == nil {
+		return nil
+	}
+
+	if len(paths) == 0 {
+		paths = ApplicationFieldPathsNested
+	}
+
+	for name, subs := range _processPaths(append(paths[:0:0], paths...)) {
+		_ = subs
+		switch name {
+		case "ids":
+
+			if v, ok := interface{}(&m.ApplicationIdentifiers).(interface{ ValidateFields(...string) error }); ok {
```

Comment:
This looks pretty dirty and I'm not sure the compiler optimizes this, but we can leave it for a future PR.

Reply:
IDEM, see https://github.com/TheThingsNetwork/lorawan-stack/pull/194#discussion_r261240612